### PR TITLE
feat: New Metrics metadata

### PR DESCRIPTION
**Pull request by chriswk**
is included in client registration and client metrics

### CODE REVIEW

**Review by sighphyre — CHANGES_REQUESTED** [review]
The code looks cool but what happened to `specVersion`?

File: config.go [hunk 1]
```diff
@@ -107,6 +108,11 @@ func WithStrategies(strategies ...strategy.Strategy) ConfigOption {
 		o.strategies = strategies
 	}
 }
+func WithStarted(startedAt time.Time) ConfigOption {
+	return func(o *configOption) {
+		o.started = &startedAt
+	}
+}
```
Comment (nunogois) [inline config.go]: Should `started` be configurable?

Comment (chriswk) [reply config.go]: Mostly for tests here

Comment (chriswk) [reply config.go]: I would've used #[cfg(test)] here, don't know what the same is in go.

Comment (chriswk) [reply config.go]: Good point, I've moved it into the test file where I used it instead.

**Comment by chriswk** [comment]
> The code looks cool but what happened to `specVersion`?

Don't you come here with your facts :1st_place_medal: . Added specVersion...

File: config.go (hunk 1 above)
Comment (nunogois) [reply config.go]: I still don't like that we're polluting our config options with something that shouldn't be user-configurable and I'm still not super convinced of its value. Maybe we can relax our tests instead?

Comment (nunogois) [reply config.go]: Something like this seems to work, but maybe there's a better way:

```go
func clientDataMatcher() func(req *http.Request, ereq *gock.Request) (bool, error) {
	defaultStrategies := []string{
		"default",
		"applicationHostname",
		"gradualRolloutRandom",
		"gradualRolloutSessionId",
		"gradualRolloutUserId",
		"remoteAddress",
		"userWithId",
		"flexibleRollout",
	}
	return func(req *http.Request, ereq *gock.Request) (bool, error) {
		var data ClientData
		err := json.NewDecoder(req.Body).Decode(&data)
		if err != nil {
			return false, err
		}

		if data.Started.IsZero() {
			return false, nil
		}

		expectedData := ClientData{
			AppName:          mockAppName,
			InstanceID:       mockInstanceId,
			SDKVersion:       fmt.Sprintf("%s:%s", clientName, clientVersion),
			Strategies:       defaultStrategies,
			Interval:         0,
			PlatformVersion:  runtime.Version(),
			PlatformName:     "go",
			YggdrasilVersion: nil,
			SpecVersion:      specVersion,
		}

		return data.AppName == expectedData.AppName &&
			data.InstanceID == expectedData.InstanceID &&
			data.SDKVersion == expectedData.SDKVersion &&
			compareStringSlices(data.Strategies, expectedData.Strategies) &&
			data.Interval == expectedData.Interval &&
			data.PlatformVersion == expectedData.PlatformVersion &&
			data.PlatformName == expectedData.PlatformName &&
			data.YggdrasilVersion == expectedData.YggdrasilVersion &&
			data.SpecVersion == expectedData.SpecVersion, nil
	}
}

func compareStringSlices(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func TestMetrics_ClientDataIncludesNewMetadata(t *testing.T) {
	assert := assert.New(t)
	defer gock.OffAll()

	gock.New(mockerServer).
		Post("/client/register").
		AddMatcher(clientDataMatcher()).
		Reply(200)

	client, err := NewClient(
		WithUrl(mockerServer),
		WithMetricsInterval(50*time.Millisecond),
		WithAppName(mockAppName),
		WithInstanceId(mockInstanceId),
	)

	assert.Nil(err, "Client should open without a problem")

	time.Sleep(100 * time.Millisecond)

	err = client.Close()

	assert.Nil(err, "Client should close without errors")

	st.Expect(t, gock.IsDone(), true)
}
```

**Review by sighphyre — APPROVED** [review]
Noice! LGTM

File: config.go (hunk 1 above)
Comment (chriswk) [reply config.go]: Thanks. I added your suggestion

Comment (chriswk) [reply config.go]: And removed started from the options

File: metrics.go [hunk 2]
```diff
@@ -73,10 +98,14 @@ type metrics struct {
 }
 
 func newMetrics(options metricsOptions, channels metricsChannels) *metrics {
+	started := time.Now()
+	if options.started != nil {
+		started = *options.started
+	}
```
Comment (nunogois) [inline metrics.go]: I believe this can be reverted now, and we can set `started: time.Now()` like before.

File: metrics_test.go [hunk 3]
```diff
@@ -426,3 +429,166 @@ func TestMetrics_ErrorCountShouldDecreaseIfSuccessful(t *testing.T) {
...
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
```
Comment (nunogois) [inline metrics_test.go:489]: These are helper functions, not tests. Maybe we should move them to the top of the file or somewhere else that's not here?

**Review by nunogois — APPROVED** [review]
LGTM - Thank you for considering my comments 😄

File: metrics_test.go (hunk 3 above)
Comment (chriswk) [reply metrics_test.go:489]: Why not here, they're used immediately after?

Comment (nunogois) [reply metrics_test.go:489]: Simply readability / convention / preference. I don't have strong feelings either way, and I understand the logic of keeping them near to where they're used, so I'm happy with keeping this.

**Pull request merged**